src/lib/variable_length_ringbuffer/VariableLengthRingbuffer.hpp

```diff
@@ -0,0 +1,115 @@
+/****************************************************************************
+ *
+ *   Copyright (C) 2023 PX4 Development Team. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in
+ *    the documentation and/or other materials provided with the
+ *    distribution.
+ * 3. Neither the name PX4 nor the names of its contributors may be
+ *    used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+
+
+#pragma once
+
+#include <stdint.h>
+#include <pthread.h>
+#include <containers/LockGuard.hpp>
+
+
+// FIFO ringbuffer implementation for packets of variable length.
+//
+// The variable length is implemented using a 4 byte header
+// containing a the length.
+//
+// The internal data is thread-safe, protected by a mutex.
+
+class VariableLengthRingbuffer
```

Comment:
Would it be possible to implement this as two classes? 
One pure ring buffer as backing memory, and then another class that does the logic with the header size etc?

This way, the logic would probably become quite a bit simpler (and thus test more asily), and we could also go around the limitation that the header has to be in a continous block.

Reply:
Ah, or you're saying the outer class would always read `sizeof(Header)` and then continue based on that. I guess that could work.

Reply:
I think I like that suggestion.

Reply:
Exactly, the inner class not knowing about header, just bytes. On read, the outer class reads `sizeof(Header)` bytes and then the amount of bytes it says in the header. On storing, it checks if there is enough available for `sizeof(Header) + payload`, then store that in the underlying ring buffer else error

Reply:
The outer class would then need to mutex, to make sure you always fully write a full packet, before attempting to read it.

Reply:
So, I tried to implement this in two layers, and there is one complication: when reading, you might pop the header but then find that the user supplied buffer is not big enough. Then you would have to scrap the packet, otherwise you would get into a garbage state. I will have to make a comment that you always have to pop with a big enough buffer!

Reply:
There is now an assert to check that it fits the outgoing buffer. This should allow to catch this problem during development. 

We can't easily ignore the case because we still have to copy the data out of the ringbuffer but we don't even have a buffer to copy it into. So it's all a bit tricky. I'd say we leave it as is with the assert in place.